Yoon

yoon627 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->

**AS-IS**
게시글,댓글은 같은 커플 코드인 사람만 조회를 할 수 있고 수정,삭제는 작성자만 할 수 있도록 구현했습니다.
게시글,댓글을 작성,수정,삭제하는 경우에 활동로그가 남도록 구현했습니다.

**TO-BE**
활동로그를 남기는 부분이나 로그인된 유저를 확인하는 부분을 AOP를 이용해 뺄 수도 있을 것 같습니다.
로그인된 유저를 매번 db에 조회해서 맞는지 검색하는 것을 줄이기 위해 현재는 username만 포함된 jwt토큰에 유저 정보를 더 넣어줘서 관련 조회 시간을 줄일 수도 있을 것 같습니다.

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [X] API 테스트 : 포스트맨을 통해 모든 API들이 정상 작동하는 것을 확인했습니다.

yoon627 committed:
fix: 토큰에 유저의 로그인 Id를 포함시키도록 수정

yoon627 committed:
fix: db의 데이터 타입과 동일하게 변경

yoon627 committed:
feat: 게시물 작성,조회,수정,삭제 구현

yoon627 committed:
fix: 이름이 헷갈려서 userId -> username으로 변경, removed_date_time 관련 column 모두 삭제, toEntity -> toDto로 이름 변경

yoon627 committed:
feat: 게시글 조회시 상세보기와 전체 불러오기 두가지로 분리, 매번 게시글 작성시 권한 확인

yoon627 committed:
feat: 댓글 작성,조회,수정,삭제 구현

yoon627 committed:
feat: 게시글,댓글 작성,수정,삭제시 활동로그를 남기는 기능 추가

mingming-mentor commented on src/main/java/com/zerobase/together/controller/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+    return ResponseEntity.ok(result);
+  }
+
+  @DeleteMapping("/delete")
```
https://restfulapi.net/resource-naming/ restful API convention이 존재합니다. 한번 참고해보시면 좋을 것 같네요

yoon627 replied:
확인해서 수정했습니다!!

mingming-mentor commented on src/main/java/com/zerobase/together/controller/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  @PutMapping("/update")
+  public ResponseEntity<?> updateComment(@RequestParam Long commentId,
```
아래 restful api convention에도 적혀있을 텐데요, 보통 requestBody가 존재하면 requestParam은 별도로 두지 않고 해당 데이터를 body로 넘깁니다

yoon627 replied:
위에서 알려주신 사이트 참고해서 수정했습니다!!

mingming-mentor commented on src/main/java/com/zerobase/together/service/CommentService.java:
```diff
@@ -0,0 +1,111 @@
+  private final PostRepository postRepository;
+  private final HistoryService historyService;
+
+  public CommentDto createComment(CommentDto request) {
```
@Transactional은 없어도 될까요?

yoon627 replied:
여러 트랜잭션이 일어나는 부분에 @transactional을 추가해줬습니다.

mingming-mentor commented on src/main/java/com/zerobase/together/service/CommentService.java:
```diff
@@ -0,0 +1,111 @@
+    if (user.getCoupleId() != this.postRepository.findById(postId).get().getCoupleId()) {
+      throw new RuntimeException("댓글 조회 권한이 없습니다.");
+    }
+    return this.commentRepository.getCommentEntitiesByPostId(postId).stream()
```
paging 처리는 없어도 되나요?

yoon627 replied:
여러개를 조회하는 부분들 모두 페이징 처리 해주엇습니다. 한페이지당 10개의 데이터가 들어가도록 했고 페이지를 가져올때 시간 역순으로 정렬시켜 가져오도록 했습니다.

mingming-mentor commented on src/main/java/com/zerobase/together/service/CommentService.java:
```diff
@@ -0,0 +1,111 @@
+      throw new RuntimeException("댓글 삭제 권한이 없습니다.");
+    }
+
+    this.commentRepository.deleteById(commentId);
```
만약 유저가 코멘트를 작성하고(id1) 이를 삭제한다면 history에 남겨진
comment 1 create 는 무용지물이 될것 같아요.

yoon627 replied:
원래는 삭제하고 생성하는 로그를 다 남겨두려고 했는데 그냥 삭제로그는 남기지 않고 게시물이나 댓글 삭제시 관련 생성,수정로그를 다 삭제하도록 바꿨습니다.

mingming-mentor replied:
현재처럼 하셔도 좋구요,
삭제한 데이터에 대해서 로그가 남고, 이를 사용자가 확인할 수 있는 구조라면 
삭제시에 실제 데이터를 삭제하지 않고 삭제여부(삭제시간)을 저장한뒤에
사용자에게 로그를 내려줄때, 삭제된 데이터의 경우 id와 삭제되었다는 status를 같이 내려줘서

로그는 보이지만 실제 사용자가 클릭해서 해당 데이터로 이동하려고 할때, client쪽에서 alert등을 띄워주는 형태로 해주셔도 좋을 것같네요

yoon627 replied:
@mingming-mentor 님 말씀해주신 방식으로 다시 변경시켜서 적용시켰습니다.
근데 그 과정에서 삭제된 게시물이나 댓글들을 제외하고 읽어오기 위해 deletedDateTime이 null인 것들을 불러오도록 조회했는데 그 방식이 적합한지 궁금합니다.

mingming-mentor replied:
네 맞습니다.

mingming-mentor commented on src/main/java/com/zerobase/together/service/CommentService.java:
```diff
@@ -0,0 +1,111 @@
+    this.historyService.createHistory(HistoryDto.builder()
+        .coupleId(user.getCoupleId())
+        .userId(user.getId())
+        .targetId(null)
```
targetId를 따로 안남기는걸까요?

yoon627 replied:
실제 view를 생각했을 때 id를 줘서 해당 게시물이나 댓글로 가는 것을 생각하고 있어서 삭제한 게시물이나 댓글에 대해서는 가지 못하도록 하려고 null을 줬었지만 일단은 삭제한 기록을 저장하는 로그를 없앴습니다.

mingming-mentor commented on src/main/java/com/zerobase/together/service/HistoryService.java:
```diff
@@ -0,0 +1,46 @@
+
+  public List<HistoryDto> readHistory() {
+    UserEntity user = getLoginUser();
+    return this.historyRepository.getHistoryEntitiesByCoupleId(user.getCoupleId()).stream()
```
실제로 view를 생각해봤을땐 history가 type, id가 넘어가는걸로는 화면을 꾸리기가 어렵지 않나 싶네요
각 테이블과 필요하다면 join을 해서 post의 경우면 제목을 comment라면 코멘트의 일부를 보여줄 수있도록 데이터를 넘겨줘야하지 않나 싶은데요
해당 API를 통해서 그려지는 화면을 생각하신게있을까요?
(FE 구현이 필요하다는 뜻이 아니라, 어떤 데이터를 api response로 내려줘야하는지에 대한 고민이 필요한 부분인것 같습니다)

yoon627 replied:
게시글이나 댓글 전체를 주면 좀 지저분해 보일거 같았는데 말씀을 듣고 보니 너무 정보가 없는 거같아서 게시글 생성 수정 시에는 게시글의 내용 10글자를 보여주고 댓글 생성시에는 댓글이 달린 게시글의 내용 10글자와 댓글의 내용 10글자를 보여주도록 바꿨습니다.

mingming-mentor reviewed (approved):
코멘트 드린 부분들이 단순한 부분이 아니라 고민지점이 많은 포인트들인것 같아요. 
한번 고민해보시면 좋을 것 같습니다

yoon627 committed:
Fix:
1. rest api convention 준수하도록 uri,파라미터 전달방식 변경
2. 여러개를 조회할시 페이징 구현
3. 삭제 history 남기지 않도록 변경
4. 게시물이나 댓글 삭제시 관련 생성,수정 history까지 삭제되도록 변경

yoon627 commented:
@mingming-mentor 님
알려주신 부분들 모두 수정했습니다. 확인부탁드립니다!!

mingming-mentor reviewed (approved):
수고 많으셨습니다

yoon627 committed:
fix: deletedDateTime을 추가하여 삭제 히스토리를 다시 남기도록 수정
